modules/javafx.graphics/src/main/java/com/sun/glass/ui/win/WinWindow.java

```diff
@@ -315,10 +325,100 @@ void setDeferredClosing(boolean dc) {
 
     @Override public void close() {
         if (!deferredClosing) {
+            if (windowControlsOverlay != null) {
+                windowControlsOverlay.dispose();
+            }
+
             super.close();
         } else {
             closingRequested = true;
             setVisible(false);
         }
     }
+
+    private WindowControlsOverlay windowControlsOverlay;
+
+    @Override
+    public WindowControlsOverlay getWindowOverlay() {
+        if (windowControlsOverlay == null && isExtendedWindow()) {
+            var url = getClass().getResource(WINDOW_DECORATION_STYLESHEET);
+            if (url == null) {
+                throw new RuntimeException("Resource not found: " + WINDOW_DECORATION_STYLESHEET);
+            }
+
+            windowControlsOverlay = new WindowControlsOverlay(StringConstant.valueOf(url.toExternalForm()));
+            windowOverlayMetrics.bind(windowControlsOverlay.metricsProperty());
+        }
+
+        return windowControlsOverlay;
+    }
+
+    @Override
+    public NonClientHandler getNonClientHandler() {
+        var overlay = getWindowOverlay();
+        if (overlay == null) {
+            return null;
+        }
+
+        return (type, button, x, y, xAbs, yAbs, clickCount) -> {
+            double wx = x / platformScaleX;
+            double wy = y / platformScaleY;
+
+            // Give the window button overlay the first chance to handle the event.
+            return overlay.handleMouseEvent(type, button, wx, wy);
+        };
+    }
+
+    /**
+     * Opens a system menu at the specified coordinates.
+     *
+     * @param x the X coordinate in physical pixels
+     * @param y the Y coordinate in physical pixels
+     */
+    public void showSystemMenu(int x, int y) {
+        _showSystemMenu(getRawHandle(), x, y);
+    }
+
+    /**
+     * Classifies the window region at the specified physical coordinate.
+     * <p>
+     * This method is called from native code.
```

Comment:
Yes, but if we do that, then we should do it everywhere it matters. The native toolkits are not multi-threaded, so all native upcalls happen on the JavaFX application thread. We should probably do this in another PR, and discuss how to do it uniformly (javadoc, annotation, etc).

Reply:
I think it would make sense to start doing this in the new code, as this helps to show the connection.  Only if as a courtesy to the developers who have to read the code years later.

Reply:
Maybe so, but let's not discuss this in this PR. It's probably best to have a discussion on the mailing list first, and come up with a consistent way of documenting native upcalls.